File: src/main/java/net/adventuregame/water/WaterShader.java
package net.adventuregame.water;

import com.chidozie.core.shaders.ShaderProgram;
import net.adventuregame.entities.Camera;
import net.adventuregame.entities.Light;
import net.adventuregame.toolbox.Maths;
import org.joml.Matrix4f;

public class WaterShader extends ShaderProgram {

    private static final String VERTEX_FILE = "water/waterVertexShader.vert";
    private static final String FRAGMENT_FILE = "water/waterFragmentShader.frag";

    private int location_modelMatrix;
    private int location_viewMatrix;
    private int location_projectionMatrix;
    private int location_reflectionTexture;
    private int location_refractionTexture;
    private int location_dudvMap;
    private int location_moveFactor;
    private int location_cameraPosition;
    private int location_normalMap;
    private int location_lightColour;
    private int location_lightPosition;
    private int location_depthMap;


    public WaterShader() {
        super(VERTEX_FILE, FRAGMENT_FILE);
    }


    @Override
    protected void bindAttributes() {
        bindAttribute(0, "position");
    }


    @Override
    protected void getAllUniformLocations() {
        location_projectionMatrix = getUniformLocation("projectionMatrix");
        location_viewMatrix = getUniformLocation("viewMatrix");
        location_modelMatrix = getUniformLocation("modelMatrix");
        location_reflectionTexture = getUniformLocation("reflectionTexture");
        location_refractionTexture = getUniformLocation("refractionTexture");
        location_dudvMap = getUniformLocation("dudvMap");
        location_moveFactor = getUniformLocation("moveFactor");
        location_cameraPosition = getUniformLocation("cameraPosition");
        location_normalMap = getUniformLocation("normalMap");
        location_lightColour = getUniformLocation("lightColour");
        location_lightPosition = getUniformLocation("lightPosition");
        location_depthMap = getUniformLocation("depthMap");
    }

    public void connectTextureUnits() {
        loadInt(location_reflectionTexture, 0);
        loadInt(location_refractionTexture, 1);
        loadInt(location_dudvMap, 2);
        loadInt(location_normalMap, 3);
        loadInt(location_depthMap, 4);
    }

    public void loadLight(Light sun) {
        loadVector3f(location_lightColour, sun.getColour());
        loadVector3f(location_lightPosition, sun.getPosition());
    }

    public void loadMoveFactor(float factor) {
        loadFloat(location_moveFactor, factor);
    }


    public void loadProjectionMatrix(Matrix4f projection) {
        loadMatrix(location_projectionMatrix, projection);
    }

    public void loadViewMatrix(Camera camera){
        Matrix4f viewMatrix = Maths.createViewMatrix(camera);
        loadMatrix(location_viewMatrix, viewMatrix);
        loadVector3f(location_cameraPosition, camera.getPosition());
    }


    public void loadModelMatrix(Matrix4f modelMatrix){
        loadMatrix(location_modelMatrix, modelMatrix);
    }

}

File: src/main/kotlin/net/adventuregame/water/WaterShader.kt
package net.adventuregame.water

import com.chidozie.core.shaders.ShaderProgram
import net.adventuregame.entity.Camera
import net.adventuregame.entity.Light
import net.adventuregame.toolbox.Maths
import org.joml.Matrix4f

class WaterShader : ShaderProgram(VERTEX_FILE, FRAGMENT_FILE) {
    private var location_modelMatrix = 0
    private var location_viewMatrix = 0
    private var location_projectionMatrix = 0
    private var location_reflectionTexture = 0
    private var location_refractionTexture = 0
    private var location_dudvMap = 0
    private var location_moveFactor = 0
    private var location_cameraPosition = 0
    private var location_normalMap = 0
    private var location_lightColour = 0
    private var location_lightPosition = 0
    private var location_depthMap = 0


    override fun bindAttributes() {
        bindAttribute(0, "position")
    }


    override fun getAllUniformLocations() {
        location_projectionMatrix = getUniformLocation("projectionMatrix")
        location_viewMatrix = getUniformLocation("viewMatrix")
        location_modelMatrix = getUniformLocation("modelMatrix")
        location_reflectionTexture = getUniformLocation("reflectionTexture")
        location_refractionTexture = getUniformLocation("refractionTexture")
        location_dudvMap = getUniformLocation("dudvMap")
        location_moveFactor = getUniformLocation("moveFactor")
        location_cameraPosition = getUniformLocation("cameraPosition")
        location_normalMap = getUniformLocation("normalMap")
        location_lightColour = getUniformLocation("lightColour")
        location_lightPosition = getUniformLocation("lightPosition")
        location_depthMap = getUniformLocation("depthMap")
    }

    fun connectTextureUnits() {
        loadInt(location_reflectionTexture, 0)
        loadInt(location_refractionTexture, 1)
        loadInt(location_dudvMap, 2)
        loadInt(location_normalMap, 3)
        loadInt(location_depthMap, 4)
    }

    fun loadLight(sun: Light) {
        loadVector3f(location_lightColour, sun.colour!!)
        loadVector3f(location_lightPosition, sun.position!!)
    }

    fun loadMoveFactor(factor: Float) {
        loadFloat(location_moveFactor, factor)
    }


    fun loadProjectionMatrix(projection: Matrix4f) {
        loadMatrix(location_projectionMatrix, projection)
    }

    fun loadViewMatrix(camera: Camera) {
        val viewMatrix = Maths.createViewMatrix(camera)
        loadMatrix(location_viewMatrix, viewMatrix)
        loadVector3f(location_cameraPosition, camera.position)
    }


    fun loadModelMatrix(modelMatrix: Matrix4f) {
        loadMatrix(location_modelMatrix, modelMatrix)
    }

    companion object {
        private const val VERTEX_FILE = "water/waterVertexShader.vert"
        private const val FRAGMENT_FILE = "water/waterFragmentShader.frag"
    }
}
